### PR TITLE
WIP Tweak `31DA` for ClimaRad

### PR DESCRIPTION
**Pull request by silverailscolo**
Complex retrieval parsing from 2 codes for Ventura + what's already in place

### CODE REVIEW

File: src/ramses_tx/parsers.py [hunk 1]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
+    if msg.len == 30 and payload[30:38] == "BE09001F":  # ClimaRad VenturaV1x 2021
+        assert payload[12:14] == "EF", f"Ventura 31DA 12: {payload[12:14]}"
+        assert payload[18:22] == "7FFF", f"Ventura 31DA 18: {payload[18:22]}"
+        assert payload[38:40] == payload[40:42], f"Ventura 31DA twin {payload[40:42]}"
+        assert payload[42:60] == "000000008500850000"
+        return {
+            "zone": f"0x{payload[0:2]}",  # zone
+            **parse_air_quality(
+                payload[2:6]
+            ),  #  N/A in VenturaV1x (reports CO2 in 12A0) but must include for Mypy??
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: If the value is `None`, then it remains in the payload dict as such.

Generally, include this information only in **tests/tests/parsers/code_31da.log**:
> `#  N/A in VenturaV1x (reports CO2 in 12A0)`

File: src/ramses_tx/parsers.py [hunk 2]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
-        **parse_capabilities(payload[30:34]),
-        **parse_bypass_position(payload[34:36]),  # 22F7-ish
-        **parse_supply_fan_speed(payload[40:42]),
-        **parse_remaining_mins(payload[42:46]),  # mins, ~22F3[2:6]
-        **parse_post_heater(payload[46:48]),
-        **parse_pre_heater(payload[48:50]),
-        **parse_supply_flow(payload[50:54]),  # NOTE: is supply, not exhaust
-        **parse_exhaust_flow(payload[54:58]),  # NOTE: order switched from others
-    }
+    if msg.len == 30 and payload[30:38] == "BE09001F":  # ClimaRad VenturaV1x 2021
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: Why should **ClimaRad** have the benefit of its own sub-parser? I'm not against the idea in principle, but it must be justified.

I think remove this, and use the existing parser.

BTW, you should generally:
```
if value:
    return A
return B
```
... instead of:
```
if value:
    return A
else:
    return B
```

File: src/ramses_tx/typed_dicts.py [hunk 3]
```diff
@@ -200,6 +200,22 @@ class _313f(TypedDict):
...
+
+class _31da_v(TypedDict):  # v for ClimaRad Ventura
+    zone: str | None
+    air_quality: AirQuality
+    air_quality_basis: NotRequired[str]
+    SZ_DEWPOINT_TEMP: float | None
+    _unknown_1: str | None
+    SZ_INDOOR_TEMP: float | None
+    SZ_EXHAUST_TEMP: float | None
+    SZ_BYPASS_POSITION: float | None
```
Comment (zxdavb) [inline src/ramses_tx/typed_dicts.py]: Nope: use the existing typed dict.

File: src/ramses_tx/parsers.py [hunk 4]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
+        assert payload[12:14] == "EF", f"Ventura 31DA 12: {payload[12:14]}"
+        assert payload[18:22] == "7FFF", f"Ventura 31DA 18: {payload[18:22]}"
+        assert payload[38:40] == payload[40:42], f"Ventura 31DA twin {payload[40:42]}"
+        assert payload[42:60] == "000000008500850000"
+        return {
+            "zone": f"0x{payload[0:2]}",  # zone
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: Please rely up on `hvac_id` instead.

Does it ever have a non-zero value?

File: src/ramses_tx/parsers.py [hunk 5]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
+            "_unknown_1": f"0x{payload[10:12]}",  # 0x00|08|3E|52|7F|9A|AF|D3 = ?
+            **parse_indoor_temp(payload[22:26]),
+            **parse_exhaust_temp(payload[26:30]),
+            **parse_bypass_position(payload[38:40]),  # 0x00|14 TODO confirm in summer
+        }  # type: ignore[return-value]
+    else:
+        return {
+            **parse_exhaust_fan_speed(
+                payload[38:40]
+            ),  # indeed in 31D9[4:6] for some, like Vasco D60
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: Put such commentary in **code_31da.log** please (code will be unreadable if all such comments, for all such devices, where put in the codebase):
> `# indeed in 31D9[4:6] for some, like Vasco D60`

File: src/ramses_tx/parsers.py [hunk 6]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
+            **parse_fan_info(payload[36:38]),  # 22F3-ish
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: With your packets, this this throws:
```
AssertionError: invalid fan_info: 31
```
... so remove the corresponding assert from **helpers.py**:
```
assert int(value, 16) & 0x1F <= 0x19, f"invalid fan_info: {int(value, 16) & 0x1F}"
```
... and put a comment about this in **code_31da.log**.

File: src/ramses_tx/parsers.py [hunk 7]
```diff
@@ -2153,27 +2154,46 @@ def parser_31d9(payload: str, msg: Message) -> dict[str, Any]:
...
+        assert payload[18:22] == "7FFF", f"Ventura 31DA 18: {payload[18:22]}"
+        assert payload[38:40] == payload[40:42], f"Ventura 31DA twin {payload[40:42]}"
+        assert payload[42:60] == "000000008500850000"
+        return {
+            "zone": f"0x{payload[0:2]}",  # zone
+            **parse_air_quality(
+                payload[2:6]
+            ),  #  N/A in VenturaV1x (reports CO2 in 12A0) but must include for Mypy??
+            SZ_DEWPOINT_TEMP: hex_to_temp(payload[6:10]),  # confirmed
+            "_unknown_1": f"0x{payload[10:12]}",  # 0x00|08|3E|52|7F|9A|AF|D3 = ?
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: Put a comment about this in **code_31da.log**.

I am not keen on putting exceptions-to-the-rule to individual systems... Particularly if we don't have an alternative.

If it is not parsing correctly, we should seek to understand why, and I'd leave it as is until then.

Do you have a 24-48h packet log with the `31DA` and the corresponding `12A0` packets in it for me to review?

File: src/ramses_tx/typed_dicts.py [hunk 8]
```diff
@@ -282,9 +298,9 @@ class FaultLogEntry(TypedDict):  # NOTE: not identical to _0418
 
 
 # These are from 31DA...
-class AirQuality(TypedDict):
-    air_quality: float | None
-    air_quality_basis: NotRequired[str]
+# class AirQuality(TypedDict): # moved before 31DA
+#    air_quality: float | None
+#    air_quality_basis: NotRequired[str]
```
Comment (zxdavb) [inline src/ramses_tx/typed_dicts.py]: I will move them all, I think, in a PR - move may have a tricky rebase, but I think is the best way.

Comment (zxdavb) [reply src/ramses_tx/typed_dicts.py]: Generally, do not leave in commented-out code.

(I do it, but it is a bad habit)

**Review by zxdavb — CHANGES_REQUESTED** [review]

File: src/ramses_tx/typed_dicts.py (hunk 8 above)
Comment (silverailscolo) [reply src/ramses_tx/typed_dicts.py]: Saw the PR, thanks. Will sort out the merge. Commented code was only left for now to mark so we could discuss it.

File: src/ramses_tx/parsers.py (hunk 4 above)
Comment (silverailscolo) [reply src/ramses_tx/parsers.py]: In my house `payload[:2]` aka `hvac_id` is always `00` but these units can be attached to a central control system using Modbus, which might require different `hvac_id`s. Will rename this part now that I got it.

File: src/ramses_tx/parsers.py (hunk 7 above)
Comment (silverailscolo) [reply src/ramses_tx/parsers.py]: Here we go:
[12A0 ClimaRad silverailscolo.log](https://github.com/user-attachments/files/17441209/12A0.ClimaRad.silverailscolo.log)
[31DA ClimaRad silverailscolo.log](https://github.com/user-attachments/files/17441210/31DA.ClimaRad.silverailscolo.log)

Comment (zxdavb) [reply src/ramses_tx/parsers.py]: Sorry - I wasn't clear - provide me a 24h (or, better, a 48h) log that has the `12A0` and the `31DA` packet in it, *as it was captured* - do not provide me a filtered packet log.

Comment (silverailscolo) [reply src/ramses_tx/parsers.py]: This the ramses_esp packet log for Oct 18:
[packetlog20241017.txt](https://github.com/user-attachments/files/17442764/packetlog20241017.txt)
(a lot of RQs)

**Comment by silverailscolo** [comment]
Contents match enough to use the existing parser, as you requested.

But I am eager to know - as in `12A0` - how we let Home Assistant understand the results for a specific system. Where does the translation from ramses_rf parse results (dict) to ramses_cc entities like "sensors" happen?

**Comment by zxdavb** [comment]
> But I am eager to know - as in `12A0` - how we let Home Assistant understand the results for a specific system. Where does the translation from ramses_rf parse results (dict) to ramses_cc entities like "sensors" happen?

This is a bit of a mess. 

Essentially,  there are 4 layers of note:
 - **ramses_esp** -- the dongle operating at the USB/MQTT <--> RF layer
 - **ramses_tx** -- the 'lower' half of this repo
 - **ramses_rf** -- the 'upper' half of this repo
 - **ramses_cc** -- the RAMSES RF custom control for HA
 
**ramses_tx** converts payloads to JSON and send them up to **ramses_rf**, which has a concept of things like FANs (ventilation units), REMs (remotes), CO2s (CO2 sensors), etc.
 
It is up to **ramses_rf** to tell you the (e.g.) current indoor humidity of a FAN. It should do that by using the latest available data, which could be either:
 - `12A0` packet
 - `31DA` packet
 
 It _should_ use the most recent packet for this...
 ... but probably doesn't.

File: src/ramses_tx/parsers.py [hunk 9]
```diff
@@ -2170,7 +2170,9 @@ def parser_31da(payload: str, msg: Message) -> PayDictT._31DA:
         **parse_outdoor_temp(payload[26:30]),  # 1290?
         **parse_capabilities(payload[30:34]),
         **parse_bypass_position(payload[34:36]),  # 22F7-ish
-        **parse_supply_fan_speed(payload[40:42]),
+        **parse_supply_fan_speed(
+            payload[40:42]
+        ),  # VenturaV1x: 0x00|08|14|28|7E repeated in [40:42]
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py]: Generally, I now want these comments in **tests/tests/parsers/code_*.log**, otherwise the code base gets messy, full of edge-case info.

**Review by zxdavb — CHANGES_REQUESTED** [review]

**Comment by silverailscolo** [comment]
Now passing after I pulled the fix in `transport.py` PR from master.

> It should use the most recent packet for this...
> ... but probably doesn't.

PRs #176 up to #183 fix this in `ramses_rf/device/hvac.py`
My complete branch running happily on local HA showing updates in UI.

File: src/ramses_tx/helpers.py [hunk 10]
```diff
@@ -721,8 +721,8 @@ def parse_fan_info(value: HexStr2) -> PayDictT.FAN_INFO:
     if not isinstance(value, str) or len(value) != 2:
         raise ValueError(f"Invalid value: {value}, is not a 2-char hex string")
 
-    # if value == "EF":  # TODO: Not implemented???
-    #     return {SZ_FAN_INFO: None}
+    if value in "1F, EF, FF":  # 1F constant in HRU which has actual fan_info in 31D9
+        return {SZ_FAN_INFO: "", "_unknown_fan_info_flags": []}
```
Comment (zxdavb) [inline src/ramses_tx/helpers.py]: I believe EF means None, or N/A - FF may be similar.

1F is completely different - it means _something_.

File: src/ramses_tx/ramses.py [hunk 11]
```diff
@@ -1339,29 +1339,29 @@
...
+    0x0B: "1 temporary override",  # timer
+    0x0C: "2 temporary override",  # timer
+    0x0D: "3 temporary override",  # timer/boost? (timer 1, 2, 3)
+    0x0E: "4 temporary override",
+    0x0F: "5 temporary override",
+    0x10: "6 temporary override",
+    0x11: "7 temporary override",
+    0x12: "8 temporary override",
+    0x13: "9 temporary override",
+    0x14: "10 temporary override",
```
Comment (zxdavb) [inline src/ramses_tx/ramses.py]: please justify changing these - the old values are the gold standard

**Review by zxdavb — CHANGES_REQUESTED** [review]

File: src/ramses_tx/ramses.py (hunk 11 above)
Comment (silverailscolo) [reply src/ramses_tx/ramses.py]: As in `22F4` please consider:
Passing along the word 'speed' in a non-English UI is not advised from i18n (internationalization). Can be added in UI by user if desired, harder to take it out. The label should make clear what a value means, so perhaps one would want to replace 'fan_info' in the UI with 'Speed' _in the UI_.

File: src/ramses_tx/helpers.py (hunk 10 above)
Comment (silverailscolo) [reply src/ramses_tx/helpers.py]: I have split this, passing along the `0x1F` value and filter on it instead.

File: src/ramses_tx/ramses.py (hunk 11 above)
Comment (silverailscolo) [reply src/ramses_tx/ramses.py]: Will follow suggestion to localize in ramses_cc, and leave API strings in lib as is

File: src/ramses_tx/parsers.py [hunk 12]
```diff
@@ -2198,6 +2204,26 @@ def parser_31da(payload: str, msg: Message) -> PayDictT._31DA:
...
+        result[SZ_SUPPLY_TEMP] = result[SZ_OUTDOOR_TEMP]  # replace
+        # [14:18] = is outdoor_temp in ClimaRad Ventura, not exhaust (too low)
+        # further confirmation: this is same value as 12A0[2], but less frequent
+        result[SZ_OUTDOOR_TEMP] = None
+        result[SZ_FAN_INFO] = (
+            None  # clear useless lookup to prevent showing in Climate UI
+        )
+
+    return result  # type: ignore[return-value]
+
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py:2254]: This is a can of worms - a slippery slope, and you should avoid this approach...

There are several examples of HVAC not following 'the' standard protocol, and doing their own thing...

They get away with it because only their remotes with their FANs, etc...

Handling these edge cases, if at all, should be done in a layer above ramses_rf.

Comment (zxdavb) [reply src/ramses_tx/parsers.py:2254]: For example, what happens if another vendor satisfies your if, but *is* following the standard?

Restore the parser, keep the logs - and add your commentary there.

**Review by zxdavb — CHANGES_REQUESTED** [review]

File: src/ramses_tx/parsers.py (hunk 12 above)
Comment (silverailscolo) [reply src/ramses_tx/parsers.py:2254]: Will take it out this weekend. Bit of a shame as it did what it had to.
Still hope to catch some of these messed up values before they are fed into the `hvac` entity, because a user (I) can't patch or swap the parts in the UI. And the climate controller entity ("thermostat") looks nice and - once the set speed works - is the place where I expect people will try to control their fan/heater.

File: src/ramses_tx/helpers.py [hunk 13]
```diff
@@ -736,8 +736,8 @@ def parse_fan_info(value: HexStr2) -> PayDictT.FAN_INFO:
     if not isinstance(value, str) or len(value) != 2:
         raise ValueError(f"Invalid value: {value}, is not a 2-char hex string")
 
-    # if value == "EF":  # TODO: Not implemented???
-    #     return {SZ_FAN_INFO: None}
+    if value in "EF, FF":
```
Comment (zxdavb) [inline src/ramses_tx/helpers.py]: `in ("EF", "FF")`

How did this pass CI and/or tests?!

File: src/ramses_tx/parsers.py [hunk 14]
```diff
@@ -2198,6 +2204,26 @@ def parser_31da(payload: str, msg: Message) -> PayDictT._31DA:
...
+        # [26:30] = is supply_temp in ClimaRad Ventura, not outdoor_temp
+        result[SZ_SUPPLY_TEMP] = result[SZ_OUTDOOR_TEMP]  # replace
+        # [14:18] = is outdoor_temp in ClimaRad Ventura, not exhaust (too low)
+        # further confirmation: this is same value as 12A0[2], but less frequent
+        result[SZ_OUTDOOR_TEMP] = None
+        result[SZ_FAN_INFO] = (
+            None  # clear useless lookup to prevent showing in Climate UI
+        )
+
+    return result  # type: ignore[return-value]
```
Comment (zxdavb) [inline src/ramses_tx/parsers.py:2253]: this is too much complexity for the base parser - needs to be in a higher-order class?

You're making promises here that are going to sting you later on - the code is unmaintainable as it is.

**Review by zxdavb — CHANGES_REQUESTED** [review]

File: src/ramses_tx/helpers.py (hunk 13 above)
Comment (silverailscolo) [reply src/ramses_tx/helpers.py]: Compares strings, but that isn't what should be used here, fixed. Thanks for your close reading

File: tests/tests/parsers/code_22f2.log [hunk 15]
```diff
@@ -14,7 +14,7 @@
...
+# 2024-10-18T14:51:12.933458 ...  I --- 32:022222 --:------ 32:022222 31DA 030 00EF007FFFEFEF087008210841086EA800EF0232320000EF0004C6027E00  # {'hvac_id': '00', 'exhaust_fan_speed': 0.25, 'fan_info': '2 (medium)', '_unknown_fan_info_flags': [0, 0, 0], 'air_quality': None, 'co2_level': None, 'indoor_humidity': None, 'outdoor_humidity': None, 'exhaust_temp': 21.6, 'supply_temp': 20.81, 'indoor_temp': 21.13, 'outdoor_temp': 21.58, 'speed_capabilities': ['off', 'timer', 'auto'], 'bypass_position': None, 'supply_fan_speed': 0.25, 'remaining_mins': 0, 'post_heat': None, 'pre_heat': 0.0, 'supply_flow': 12.22, 'exhaust_flow': 6.38}
```
Comment (silverailscolo) [inline tests/tests/parsers/code_22f2.log]: Should include word 'speed'

**Comment by silverailscolo** [comment]
Why is the client.py test broken by these changes? 
- Timing issue?
- Wrong thread?

Errors disappeared after pulling master.

```
2025-08-02T21:20:34.8207241Z   [90m [39;49;00m }[90m[39;49;00m
2025-08-02T21:20:34.8209330Z [31mFAILED[0m tests/tests/test_cli_utility.py::[1mtest_client_basic[monitor][0m - AssertionError: assert {'config': {'.../dev/ttyUSB0'} == {'config': {'.../dev/ttyUSB0'}
2025-08-02T21:20:34.8210420Z   
2025-08-02T21:20:34.8210737Z   Omitting 2 identical items, use -vv to show
2025-08-02T21:20:34.8211177Z   Differing items:
2025-08-02T21:20:34.8214359Z   [0m{[33m'[39;49;00m[33mconfig[39;49;00m[33m'[39;49;00m: {[33m'[39;49;00m[33menable_eavesdrop[39;49;00m[33m'[39;49;00m: [94mFalse[39;49;00m, [33m'[39;49;00m[33mevofw_flag[39;49;00m[33m'[39;49;00m: [94mNone[39;49;00m, [33m'[39;49;00m[33mreduce_processing[39;49;00m[33m'[39;49;00m: [94m0[39;49;00m}}[90m[39;49;00m != [0m{[33m'[39;49;00m[33mconfig[39;49;00m[33m'[39;49;00m: {[33m'[39;49;00m[33mdisable_discovery[39;49;00m[33m'[39;49;00m: [94mFalse[39;49;00m, [33m'[39;49;00m[33mevofw_flag[39;49;00m[33m'[39;49;00m: [94mNone[39;49;00m, [33m'[39;49;00m[33mreduce_processing[39;49;00m[33m'[39;49;00m: [94m0[39;49;00m}}[90m[39;49;00m
2025-08-02T21:20:34.8217034Z   
2025-08-02T21:20:34.8217349Z   Full diff:
2025-08-02T21:20:34.8217677Z   [0m[90m [39;49;00m {[90m[39;49;00m
2025-08-02T21:20:34.8218078Z   [90m [39;49;00m     'config': {[90m[39;49;00m
2025-08-02T21:20:34.8218569Z   [91m-         'disable_discovery': False,[39;49;00m[90m[39;49;00m
2025-08-02T21:20:34.8219032Z   ?          ^^^     ^^^^^   ^[90m[39;49;00m
2025-08-02T21:20:34.8219495Z   [92m+         'enable_eavesdrop': False,[39;49;00m[90m[39;49;00m
2025-08-02T21:20:34.8219937Z   ?          ^^     ^^  ++ ^^[90m[39;49;00m
2025-08-02T21:20:34.8220375Z   [90m [39;49;00m         'evofw_flag': None,[90m[39;49;00m
2025-08-02T21:20:34.8220859Z   [90m [39;49;00m         'reduce_processing': 0,[90m[39;49;00m
2025-08-02T21:20:34.8221555Z   [90m [39;49;00m     },[90m[39;49;00m
2025-08-02T21:20:34.8221975Z   [90m [39;49;00m     'packet_log': None,[90m[39;49;00m
2025-08-02T21:20:34.8222487Z   [90m [39;49;00m     'serial_port': '/dev/ttyUSB0',[90m[39;49;00m
2025-08-02T21:20:34.8222921Z   [90m [39;49;00m }[90m[39;49;00m
2025-08-02T21:20:34.8223955Z [31mFAILED[0m tests/tests/test_cli_utility.py::[1mtest_client_basic[listen ][0m - AssertionError: assert {'config': {'.../dev/ttyUSB0'} == {'config': {'.../dev/ttyUSB0'}
2025-08-02T21:20:34.8224775Z   
```
